sycl/doc/SYCLPluginInterface.md

```diff
@@ -41,12 +40,45 @@ once before any actual offload is attempted.
 
 ### Plugin discovery
 
-Plugins are physically dynamic libraries stored somewhere in the system where
-the SYCL runtime runs. TBD - design and describe the process in details.
+Plugins are physically dynamic libraries or shared objects.
+The process to discover plugins will follow the following guidelines.
+
+The SYCL Runtime will search for plugins at env LD_LIBRARY_PATH location on
+Linux and env LIB on Windows, with names in the format of libpiXXX.so and
+query them. An extension to the search mechanism is to use a configuration file
```

Comment:
The main motivation is to have less configuration needed from the end user, and thus being more user-friendly. In the worst case, the pre-defined location of such configuration (e.g. where the SYCL runtime library is) could be on a drive with no write access. We could change the matching pattern to be more complex (e.g. libsyclpi*.so) to get less false positives.

Reply:
Do I understand correctly that there will be  OpenCL PI plugin which loads libOpenCL.so?

Reply:
> Do I understand correctly that there will be OpenCL PI plugin which loads libOpenCL.so?

The libpi_opencl.so would just be statically linked against the libOpenCL.so, and yes, later shall be loaded at the time the libpi_opencl.so is loaded by the SYCL RT.

Reply:
Yes, this is what it would be in more details:
A configuration file with some predefined (TBD) location inside SYCL RT will be listing the names of the PI plugin libraries to search for and load. The search is made in paths given by LD_LIBRARY_PATH. Also, to address the issue of read-only configuration file, there will be an environment variable added (e.g. SYCL_PI_CONFIG) to provide a user-specified config file, which, if given, would override the predefined location inside SYCL RT.

Reply:
It would be good if it would not be necessary to add plugin directories to `LD_LIBRARY_PATH` (on Linux, nor `PATH` on Windows) for the runtime to use the plugins packaged with it.

## Linux

On Linux an [`DT_RUNPATH`](https://linux.die.net/man/8/ld.so) can be embedded into the SYCL library for [dlopen](https://linux.die.net/man/3/dlopen) to look relative to the location of the SYCL library for plugin libraries. The `LD_LIBRARY_PATH` env var will override where to look first.

## Windows

On Windows locate the directory of the SYCL library and then use [`SetDllDirectory`](https://docs.microsoft.com/en-gb/windows/win32/api/winbase/nf-winbase-setdlldirectorya) (although this turns "SafeDllSearchmode" off...) to where the related plugin directory is located, or, use [`LoadLibraryEx`](https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibraryexa) instead.

The difference to Linux is that [`LoadLibraryA`](https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibrarya) and `LoadLibraryEx` prefer the path set above over the `PATH` env var.

The details of packaging an app with SYCL for the Windows Store will need extra consideration.

Reply:
I agree that it would be good if shipped PI plugins could be auto-discovered in some predefined locations relative to the installed SYCL RT.

Reply:
I have updated the doc based on the comments. Please let me know if you see any further issues.
@smaslov-intel : Why are we not taking full paths in the config file instead of just names to avoid looking at folders?
When the user has issues with read-only access, they can use the Env variable and add more plugins with the full path to the plugin.